kolibri/auth/models.py

```diff
@@ -1,497 +1,428 @@
+
+    class Meta:
+        abstract = True
+
+    def clean_fields(self, *args, **kwargs):
+        # ensure that we have, or can infer, a dataset for the model instance
+        self.ensure_dataset()
+        super(AbstractFacilityDataModel, self).clean_fields(*args, **kwargs)
+
+    def save(self, *args, **kwargs):
+
+        # before saving, ensure we have a dataset, and convert any validation errors into integrity errors,
+        # since by this point the `clean_fields` method should already have prevented this situation from arising
+        try:
+            self.ensure_dataset()
+        except KolibriValidationError as e:
+            raise IntegrityError(str(e))
+
+        super(AbstractFacilityDataModel, self).save(*args, **kwargs)
+
+    def ensure_dataset(self):
+        """
+        If no dataset has yet been specified, try to infer it. If a dataset has already been specified, to prevent
+        inconsistencies, make sure it matches the inferred dataset, otherwise raise a KolibriValidationError.
+        If we have no dataset and it can't be inferred, we raise a KolibriValidationError exception as well.
+        """
+        inferred_dataset = self.infer_dataset()
+        if self.dataset_id:
+            # make sure currently stored dataset matches inferred dataset, if any
+            if inferred_dataset and inferred_dataset != self.dataset:
+                raise KolibriValidationError("This model is not associated with the correct FacilityDataset.")
+        else:
+            # use the inferred dataset, if there is one, otherwise throw an error
+            if inferred_dataset:
+                self.dataset = inferred_dataset
+            else:
+                raise KolibriValidationError("FacilityDataset ('dataset') not provided, and could not be inferred.")
+
+    def infer_dataset(self):
+        """
+        This method is used by `ensure_dataset` to "infer" which dataset should be associated with this instance.
+        It should be overridden in any subclass of AbstractFacilityDataModel, to define a model-specific inference.
+        """
+        raise NotImplementedError("Subclasses of AbstractFacilityDataModel must override the `infer_dataset` method.")
+
+
+class KolibriAbstractBaseUser(AbstractBaseUser):
     """
     Our custom user type, derived from AbstractBaseUser as described in the Django docs.
-    Draws liberally from django.contrib.auth.AbstractUser, except we remove some fields we don't care about, like
-    email, and we don't use the PermissionsMixin.
-    Encapsulates both FacilityUsers and DeviceOwners, which are proxy models.
+    Draws liberally from django.contrib.auth.AbstractUser, except we exclude some fields
+    we don't care about, like email.
 
-    You should prefer to use the proxy models for this class where possible.
+    This model is an abstract model, and is inherited by both FacilityUser and DeviceOwner.
     """
+
+    class Meta:
+        abstract = True
```

Comment:
As discussed, the base Kolibri user is now abstract (and DeviceOwner/FacilityUser are no longer proxies).